### PR TITLE
[caplin] incremental improvements to p2p layer

### PR DESCRIPTION
**Pull request by elee1766**
we update observability in the p2p layer for handlers, and also properly encode error codes, close streams.

### CODE REVIEW

**Review by Giulio2002 — APPROVED** [review]
LGTM

File: cl/beacon/router.go [hunk 1]
```diff
@@ -0,0 +1,79 @@
...
+	mux chi.Router
+}
+
+func (a *ApiHandler) init() {
+	r := chi.NewRouter()
+	a.mux = r
+
+	// This is the set of apis for validation + otterscan
+	// otterscn specific ones are commented as such
+	r.Route("/eth", func(r chi.Router) {
```
Comment (AskAlexSharov) [inline cl/beacon/router.go:21]: REST api? Is it specified somewhere?

File: cl/sentinel/peers/peer.go [hunk 2]
```diff
@@ -0,0 +1,146 @@
...
+	// acts as the mutex for making requests. channel used to avoid use of TryLock
+	working chan struct{}
+	// backref to the manager that owns this peer
+	m *Manager
+}
+
+func (p *Peer) do(fn func(p *Peer)) {
+	p.mu.Lock()
+	fn(p)
+	p.mu.Unlock()
```
Comment (AskAlexSharov) [inline cl/sentinel/peers/peer.go:40]: need defer, because fn can panic

**Review by AskAlexSharov — CHANGES_REQUESTED** [review]

File: cl/beacon/router.go [hunk 3]
```diff
@@ -0,0 +1,79 @@
+package beacon
```
Comment (Giulio2002) [inline cl/beacon/router.go:1]: I am fine ignoring this for the time being

File: cl/phase1/stages/stage_fork_choice.go [hunk 4]
```diff
@@ -32,7 +32,8 @@ type StageForkChoiceCfg struct {
 	forkChoice      *forkchoice.ForkChoiceStore
 }
 
-const minPeersForDownload = 3
+const minPeersForDownload = 2
```
Comment (Giulio2002) [inline cl/phase1/stages/stage_fork_choice.go:35]: so downloading blocks is different than starting the sync? why would that be?

File: cl/phase1/stages/stage_fork_choice.go [hunk 5]
```diff
@@ -153,8 +155,17 @@ MainLoop:
 			continue
 		}
 		waitWhenNotEnoughPeers := 5 * time.Second
-		if peersCount < minPeersForDownload {
-			log.Debug("Cannot sync up caplin, not enough peers", "have", peersCount, "needed", minPeersForDownload, "retryIn", waitWhenNotEnoughPeers)
+		if !isDownloading {
+			isDownloading = peersCount >= minPeersForSyncStart
+		}
+		if isDownloading {
```
Comment (Giulio2002) [inline cl/phase1/stages/stage_fork_choice.go:161]: I think i get it so sync starts at 4 peers but will stop at 2 peers, correct?

File: cmd/sentinel/sentinel/communication/send_request.go [hunk 6]
```diff
@@ -19,42 +17,42 @@ var NoRequestHandlers = map[string]bool{
...
-	defer respRetryTicker.Stop()
+func SendRequestRawToPeer(ctx context.Context, host host.Host, data []byte, topic string, peerId peer.ID) ([]byte, byte, error) {
 
-	stream, err := writeRequestRaw(host, ctx, data, peerId, topic)
+	nctx, cn := context.WithTimeout(ctx, 5*time.Second)
+	defer cn()
+	stream, err := writeRequestRaw(host, nctx, data, peerId, topic)
 	if err != nil {
-		return nil, false, err
+		return nil, 189, err
```
Comment (Giulio2002) [inline cmd/sentinel/sentinel/communication/send_request.go:32]: what is 189

File: cmd/sentinel/sentinel/communication/send_request.go [hunk 7]
```diff
@@ -72,16 +70,16 @@ func writeRequestRaw(host host.Host, ctx context.Context, data []byte, peerId pe
...
-		return nil, false, fmt.Errorf("failed to read code byte peer=%s, err=%s", peerId, err)
+	_, resp.err = stream.Read(code)
+	if resp.err != nil {
+		return
 	}
-
-	message, err := io.ReadAll(stream)
-	if err != nil {
-		return nil, false, err
+	resp.code = code[0]
```
Comment (Giulio2002) [inline cmd/sentinel/sentinel/communication/send_request.go:79]: lgtm here

File: cl/sentinel/peers/peer.go (hunk 2 above)
Comment (elee1766) [reply cl/sentinel/peers/peer.go:40]: done

**Review by Giulio2002 — APPROVED** [review]

**Review by Giulio2002 — APPROVED** [review]

**Review by AskAlexSharov — APPROVED** [review]

**Pull request merged**